### PR TITLE
ARROW-8791: [Rust] Allow creation of StringDictionaryBuilder with an existing array of dictionary values

### PR DESCRIPTION
**Pull request by jhorstmann**
As mentioned in the ticket, I'm not sure this is the best way to implement this functionality, I'm open for feedback and suggestions.

### CODE REVIEW

File: rust/arrow/src/array/builder.rs [hunk 1]
```diff
@@ -1334,6 +1334,35 @@ where
             map: HashMap::new(),
         }
     }
+
+    pub fn new_with_dictionary(
```
Comment (nevi-me) [inline rust/arrow/src/array/builder.rs:1338]: may you please add a doc comment

**Review by nevi-me — COMMENTED** [review]
Thanks @jhorstmann, I'll review properly after you add a doc comment on the function. I have a question or 2, but I think they'll be addressed by documentation.

Please check the `testing` submodule, I think you need to update it to the latest commit on master (https://github.com/apache/arrow-testing/commit/bff46cad1a4b4df490f853a319e143a03b6ec233)

**Comment by jhorstmann** [comment]
While writing the docs, I was wondering what is actually the usecase for the existing `new` function taking key and value builders as argument vs. taking two capacity values. It reduces the possibility for mixing up those two capacities, but I can't think of a reason for creating a dictionary builder with something else than two freshly instantiated builders.

File: rust/arrow/src/array/builder.rs [hunk 2]
```diff
@@ -1334,6 +1334,58 @@ where
...
+    /// let mut builder = StringDictionaryBuilder::new_with_dictionary(PrimitiveBuilder::<Int16Type>::new(3), &dictionary_values).unwrap();
+    /// builder.append("def").unwrap();
+    /// builder.append_null().unwrap();
+    /// builder.append("abc").unwrap();
+    ///
+    /// let dictionary_array = builder.finish();
+    ///
+    /// let keys: Vec<Option<i16>> = dictionary_array.keys().collect();
+    ///
+    ///  assert_eq!(keys, vec![Some(2), None, Some(1)]);
```
Comment (vertexclique) [inline rust/arrow/src/array/builder.rs:1359]: assert is off by one space.

File: rust/arrow/src/array/builder.rs [hunk 3]
```diff
@@ -1334,6 +1334,58 @@ where
...
+    /// ```
+    pub fn new_with_dictionary(
+        keys_builder: PrimitiveBuilder<K>,
+        dictionary_values: &StringArray,
+    ) -> Result<Self> {
+        let mut values_builder = StringBuilder::with_capacity(
+            dictionary_values.len(),
+            dictionary_values.value_data().len(),
+        );
+        let mut map: HashMap<Box<[u8]>, K::Native> = HashMap::new();
```
Comment (vertexclique) [inline rust/arrow/src/array/builder.rs:1369]:
```suggestion
        let mut map: HashMap<Box<[u8]>, K::Native> = HashMap::with_capacity(dictionary_values.len());
```

File: rust/arrow/src/array/builder.rs [hunk 4]
```diff
@@ -1334,6 +1334,58 @@ where
...
+        for i in 0..dictionary_values.len() {
```
Comment (vertexclique) [inline rust/arrow/src/array/builder.rs:1370]: maybe better.
```suggestion
        0..dictionary_values.len()
        .for_each(|i| {
```

**Review by vertexclique — APPROVED** [review]
Ornamental nits, apart from that looks good.

File: rust/arrow/src/array/builder.rs [hunk 5]
```diff
@@ -1334,6 +1334,57 @@ where
             map: HashMap::new(),
         }
     }
+
+    /// Creates a new `StringDictionaryBuilder` from a keys builder and a dictionary initialized with the given values.
```
Comment (sunchao) [inline rust/arrow/src/array/builder.rs:1338]: nit: can we limit the line width to 90 characters?

**Review by sunchao — APPROVED** [review]
LGTM except a nit

**Review by nevi-me — APPROVED** [review]

**Comment by jhorstmann** [comment]
Hi @nevi-me and @sunchao thanks for the review, if there are no further review comments it would be awesome if we could get this merged.

**Pull request closed without merging**